iso15118/secc/states/iso15118_2_states.py

```diff
@@ -1597,6 +1611,50 @@ async def check_state():
                 CpState.D2,
             ]
 
+    def _is_charging_profile_valid(self, power_delivery_req: PowerDeliveryReq) -> bool:
+        for schedule in self.comm_session.offered_schedules:
+            if schedule.sa_schedule_tuple_id == power_delivery_req.sa_schedule_tuple_id:
+                schedule_entries = schedule.p_max_schedule.schedule_entries
+                # schedule_entries shall look like [PMaxScheduleEntry_0,
+                # PMaxScheduleEntry_1, ...]. The enumerate will pair an index with
+                # each list entry: (0, PMaxScheduleEntry_0), (1,
+                # PMaxScheduleEntry_1), ...
+                for idx, sa_profile_entry in enumerate(schedule_entries):
+                    sa_profile_entry_start = sa_profile_entry.time_interval.start
+                    sa_entry_pmax = (
+                        sa_profile_entry.p_max.value
+                        * 10**sa_profile_entry.p_max.multiplier
+                    )
+                    for (
+                        ev_profile_entry
+                    ) in power_delivery_req.charging_profile.profile_entries:
+                        try:
+                            # By getting the next entry/slot, we can know when
+                            # the current entry ends
+                            next_sa_profile_entry = schedule_entries[idx + 1]
+                            sa_profile_entry_end = (
+                                next_sa_profile_entry.time_interval.start
+                            )
+
+                        except IndexError:
+                            # As the index is out of rage, it signals this is
+                            # the final entry of the sa profile. The last entry
+                            # must have a duration field, so we can calculate
+                            # the end of the period summing the relative start
+                            # of the entry with the duration of it
+                            sa_profile_entry_end = (
+                                sa_profile_entry_start
+                                + sa_profile_entry.time_interval.duration
+                            )
+                        # fmt: off
+                        if sa_profile_entry_start <= ev_profile_entry.start <= sa_profile_entry_end:  # noqa
+                            ev_entry_pmax = (ev_profile_entry.max_power.value
+                                             * 10 ** ev_profile_entry.max_power.multiplier)  # noqa
+                            if ev_entry_pmax > sa_entry_pmax:
+                                return False
```

Comment:
So, I went over this and realized that that was not the problem. the code was breaking the loop, but it needed some more logic to work as expected. I added it and also included a more elaborated test and looks good ;)

Reply:
Yeah, I tested as well, the problem wasn't this, we should check all profile entries to find the correct range to be checked. it passed all unit tests. Thanks.

Reply:
yeah, we had to check all the entries with the logic before, but the more EV profile entries more time it would take as in the worst case scenario, if we have N EV profile entries and M SA schedule entries, we would do NxM checks all the time. And the rationale is that the EV entries we have already checked during one iteration, do not need to be reevaluated for the next one.